app/src/main/java/com/nutomic/syncthingandroid/model/Completion.java

```diff
@@ -0,0 +1,138 @@
+package com.nutomic.syncthingandroid.model;
+
+import android.util.Log;
+
+import com.nutomic.syncthingandroid.util.DefaultHashMap;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+/**
+ * This class caches remote folder and device synchronization
+ * completion indicators defined in {@link CompletionInfo#CompletionInfo}
+ * according to syncthing's REST "/completion" JSON result schema.
+ * Completion model of syncthing's web UI is completion[deviceId][folderId]
+ */
+public class Completion {
+
+    private static final String TAG = "Completion";
+
+    DefaultHashMap<String, DefaultHashMap<String, CompletionInfo>> deviceFolderMap =
+        new DefaultHashMap<String, DefaultHashMap<String, CompletionInfo>>();
+
+    DefaultHashMap<String, CompletionInfo> defaultDevice = new DefaultHashMap<String, CompletionInfo>();
+
+    /**
+     * Removes a folder from the cache model.
+     */
+    private void removeFolder(String folderId) {
+        for (String deviceId : deviceFolderMap.keySet()) {

```

Comment:
You can definately use an interator here as you are not modifying the map you are iterating over.

Reply:
Yes, but why do the work in second valid way? Java 5 allows us to direct usage of the iterator, with a for-each loop. What is the advantage? I've also in mind that an iterator does not solve the problem we are not allowed to immediately remove an element while iterating through.

Reply:
Well it gets rid of the double call to get for starters. You are not removing elements from deviceFolderMap, you are removing it from the submap for the given device.

Reply:
You can also simply use valueSet

Reply:
Oh good point :)

Reply:
As I understand getValue() will copy the object, so I would leave the get as it is in the removeFolder function. We need a reference there. To improve it, I store the reference in folderMap after the one and only get.